### PR TITLE
New saga testing framework

### PR DESCRIPTION
**Pull request by DavidBoike**

### CODE REVIEW

File: src/NServiceBus.Testing.Tests/Sagas/ExternalHandlerSimulator.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+            testableSaga.SimulateHandlerResponse<RunStep1Command, Step1Response>(cmd => new Step1Response());
+
+            var startResult = await testableSaga.Process(new StartMsg { CorrId = "abc" });
+
+            Assert.That(startResult.Context.SentMessages.Count, Is.EqualTo(1));
+            Assert.That(testableSaga.QueueLength, Is.EqualTo(1));
+            Assert.That(testableSaga.QueuePeek().Type, Is.EqualTo(typeof(Step1Response)));
+
+            var continueResult = await testableSaga.ProcessQueuedMessage();
```
Comment (andreasohlund) [inline src/NServiceBus.Testing.Tests/Sagas/ExternalHandlerSimulator.cs:23]: What is the benefit of this compared to just call Process with the response message?

Comment (DavidBoike) [reply src/NServiceBus.Testing.Tests/Sagas/ExternalHandlerSimulator.cs:23]: There is a method like that, but feels horribly inelegant given you have to also provide the saga id for the auto correlation, something I’d prefer to be inside the black box.

Comment (andreasohlund) [reply src/NServiceBus.Testing.Tests/Sagas/ExternalHandlerSimulator.cs:23]: Got it, that is clearly a benefit. It made me realize that this style of testing also tests mappings and consecutive message exchanges compared to pure "unit test one handler at a time". I like that you have the entire flow of message for a "scenario" right in front of you in the same test method.

👍

File: src/NServiceBus.Testing.Tests/Sagas/BasicSaga.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    [TestFixture]
+    public class BasicSaga
+    {
+        [Test]
+        public async Task TestBasicSaga()
+        {
+            var testableSaga = new TestableSaga<ShippingPolicy, ShippingPolicyData>();
+
+            var placeResult = await testableSaga.Process(new OrderPlaced { OrderId = "abc" });
```
Comment (andreasohlund) [inline src/NServiceBus.Testing.Tests/Sagas/BasicSaga.cs:16]: Would there be any merit in having a `.Start` overload to make things more explicit? 

Disclaimer: I know I have been on record saying that "pretty much all messages should be allowed to start a saga" but as long as we have `IAmStartedBy...` this type of feature might still make sense

File: src/NServiceBus.Testing.Tests/Sagas/ExternalHandlerSimulator.cs (hunk 1 above)
Comment (DavidBoike) [reply src/NServiceBus.Testing.Tests/Sagas/ExternalHandlerSimulator.cs:23]: I like that word `Scenario`, I had been using the word conversation but that didn't feel right. Would something with Scenario in it be a better name for this than `TestableSaga`? Though `TestableX` is kind of the pattern of everything else in here like `TestableMessageHandlerContext` and the testable saga isn't really a scenario either - your test can test a scenario, but the thing itself is analogous to a deployed saga, with its own storage and able to differentiate between messages from different conversations.

All this to say I hate naming things.

File: src/NServiceBus.Testing.Tests/Sagas/BasicSaga.cs (hunk 2 above)
Comment (DavidBoike) [reply src/NServiceBus.Testing.Tests/Sagas/BasicSaga.cs:16]: I get what you're trying to say but in the context of testing, the saga is a thing you throw a bunch of messages at and hope the right things pop out, so differentiating the methods used to throw the messages at the saga sounds odd to me.

File: src/NServiceBus.Testing/NSB.Testing.Fakes/TestingExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    /// TODO
+    /// </summary>
+    public static class TestingExtensions
+    {
+        /// <summary>
+        /// TODO
+        /// </summary>
+        /// <typeparam name="TMessageType"></typeparam>
+        /// <returns></returns>
+        public static TMessageType FindPublishedMessage<TMessageType>(this TestablePipelineContext context)
```
Comment (danielmarbach) [inline src/NServiceBus.Testing/NSB.Testing.Fakes/TestingExtensions.cs:15]: Based on the name my expectation would be that it is possible to **not** find a message and then the result would be null

**Review by danielmarbach — COMMENTED** [review]
In general I like it. One thing that I find a bit hard to grasp is that most of the time the APIs return a result that represent the state of what happened but then sometimes the testable saga remains stateful and has to be explicitly triggered by calling a method that than forwards the internal state to some new derived state. 

I would generally prefer an API where the side effects would be returned from the process methods and can be passed into the next process method to rehydrate a specific contextual state including a new message. I totally haven thought this through in-depth but here is a wild idea. 

```csharp
           var testableSaga = new TestableSaga<MySaga, MyData>();
           var startResult = await testableSaga.Process(new StartMsg { CorrId = "abc" });
           var continueResult = await  testableSaga.Process(new Step1Response(), startResult);
``` 

with send local

```csharp
            var testableSaga = new TestableSaga<MySaga, MyData>();
            var startResult = await testableSaga.Process(new StartMsg { CorrId = "abc" });
            var continueResult = await  testableSaga.Process(startResult);
```

or with timeouts

```csharp
            var testableSaga = new TestableSaga<ShippingPolicy, ShippingPolicyData>();
            var placeResult = await testableSaga.Process(new OrderPlaced { OrderId = "abc" });
            var billResult = await testableSaga.Process(new OrderBilled { OrderId = "abc" });
            var noResults = await testableSaga.Process(billResult, TimeSpan.FromMinutes(10));
            var timeoutResults = await testableSaga.Process(billResult, TimeSpan.FromMinutes(15));
```

and if you change the order of "message arriving"

```csharp
            var testableSaga = new TestableSaga<ShippingPolicy, ShippingPolicyData>();
            var billResult = await testableSaga.Process(new OrderBilled { OrderId = "abc" });
            var placeResult = await testableSaga.Process(new OrderPlaced { OrderId = "abc" });
            var noResults = await testableSaga.Process(placeResult, TimeSpan.FromMinutes(10));
            var timeoutResults = await testableSaga.Process(placeResult, TimeSpan.FromMinutes(15));
```

Furthermore if we have convenience API I would love to get the extra mile to reduce some of the trainwrecks like `result.Context....` and directly convenience methods on the strongly typed results we return. 

Additional brain farts:

Maybe by having multiple Result types we can build more convenience API on top of the result types that allows us to have proper pattern matching and maybe types to express higher level concepts. For example something like

```csharp
var processResults = await testableSaga.Process(placeResult, TimeSpan.FromMinutes(10));
var maybeTimeouts = processResults.Timeouts<TimeoutType>(); // <-- returns None or timeout state of given type if any
```

File: src/NServiceBus.Testing/NSB.Testing.Fakes/TestingExtensions.cs (hunk 3 above)
Comment (DavidBoike) [reply src/NServiceBus.Testing/NSB.Testing.Fakes/TestingExtensions.cs:13]: Changed all these to FirstOrDefault and made it clear in the xmldoc that null was returned of none found.

File: src/NServiceBus.Testing/Sagas/TestableSaga.cs [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+        /// Returns a <see cref="MessageProcessingResult"/> containing a snapshot of the saga data after the message is processed
+        /// and the <see cref="TestableMessageHandlerContext"/> containing details about context operations that occurred in the message
+        /// handler, which can be used for test assertions.
+        /// </returns>
+        public Task<MessageProcessingResult> ProcessAutoCorrelatedReply<TMessageType>(Guid sagaId, TMessageType message, TestableMessageHandlerContext context = null, IReadOnlyDictionary<string, string> messageHeaders = null)
+        {
+            Dictionary<string, string> newHeaders;
+            if (messageHeaders != null)
+            {
+                newHeaders = new Dictionary<string, string>((IDictionary<string, string>)messageHeaders);
```
[the remark this answers is not in the document]
Comment (DavidBoike) [reply src/NServiceBus.Testing/Sagas/TestableSaga.cs:131]: I think you were sleepy, that's what it's doing by creating a new dictionary, giving it the values from the read-only dictionary. You just have to cast the read-only dictionary as an IDictionary to get the Dictionary constructor to accept it.

[the remark this answers is not in the document]
Comment (danielmarbach) [reply src/NServiceBus.Testing/Sagas/TestableSaga.cs:131]: I don't think so. By contract a IReadOnlyDictionary doesn't have to implement IDictionary. Therefore the cast can be in certain cases invalid. That's why I pointed out if you declare the input parameter as readonly, all you should do is foreach through the key value pairs and write them to the new one.

[the remark this answers is not in the document]
Comment (DavidBoike) [reply src/NServiceBus.Testing/Sagas/TestableSaga.cs:131]: Ah ok, I got you. Fixed.

**Review by cquirosj — APPROVED** [review]

**Review by TravisNickels — APPROVED** [review]

File: src/NServiceBus.Testing/Sagas/SagaMapper.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Linq.Expressions;
+    using System.Reflection;
+    using System.Threading.Tasks;
+    using NServiceBus.Sagas;
+
+    class SagaMapper : IConfigureHowToFindSagaWithMessage, IConfigureHowToFindSagaWithMessageHeaders
+    {
+        static Dictionary<Type, SagaMapper> sagaMappers = new Dictionary<Type, SagaMapper>();
```
Comment (danielmarbach) [inline src/NServiceBus.Testing/Sagas/SagaMapper.cs:13]: Have we made sure this is thread safe? XUnit for example parallelizes tests I think almost by default. NUnit allows to opt in for parallelism. Then we will run into concurrent modifications of the mappers

Comment (DavidBoike) [reply src/NServiceBus.Testing/Sagas/SagaMapper.cs:13]: Argh. I'll make all the Things into ConcurrentThings tomorrow.

**Pull request merged**